server/src/controllers/grades.ts

```diff
@@ -88,3 +95,324 @@ export async function addGrades(req: Request, res: Response, next: NextFunction)
   // Close the readable stream once data reading finished.
   parser.end();
 }
+
+/**
+ * Recursively calculates the grade for a particular attainment by its formula
+ * node.
+ *
+ * TODO: Save calculated grades to the database.
+ */
+async function calculateFormulaNode(
+  formulaNode: FormulaNode,
+  presetGrades: Map<FormulaNode, number>
+): Promise<GradingResult> {
+  /*
+   * If a teacher has manually specified a grade for this attainment,
+   * the manually specified grade will be used.
+   *
+   * A grade has to be manually specified when using the 'Manual' formula
+   * and a grade may be manually specified for overriding grade calculation
+   * functions in special cases.
+   */
+  const presetGrade: number | undefined = presetGrades.get(formulaNode);
+  if (presetGrade) {
+    return {
+      status: Status.Pass,
+      grade: presetGrade
+    };
+  }
+
+  /*
+   * A teacher has not manually specified a grade for this attainment, therefore
+   * the grade will be calculated based on the given formula and the grades
+   * from this attainment's subattainments.
+   */
+
+  // Inputs for the formula of this attainment.
+  const inputs: Array<GradingInput> = [];
+
+  // First, recursively calculate the grades the subattainments.
+  for (const subFormulaNode of formulaNode.subFormulaNodes) {
+    const input: GradingInput = {
+      subResult: await calculateFormulaNode(subFormulaNode, presetGrades),
+      params: subFormulaNode.parentFormulaParams
+    };
+
+    inputs.push(input);
+  }
+
+  // Then, calculate the grade of this attainment using the formula specified
+  // for this attainment and the grades of the subattainments.
+  return await formulaNode.formulaImplementation.formulaFunction(inputs);
+}
+
+interface UserAttainmentGradeWithUser extends UserAttainmentGrade {
+  User: User
+}
+
+export async function calculateGrades(
+  req: Request,
+  res: Response
+): Promise<void> {
+  const courseId: number = Number(req.params.courseId);
+  const courseInstanceId: number = Number(req.params.instanceId);
+  validateCourseAndInstance(courseId, courseInstanceId);
+
+  /*
+   * First we need to get all the attainments in this course instance.
+   */
+
+  const attainments: Array<{
+    id: number,
+    parentId: number,
+    formula: Formula | null,
+    parentFormulaParams: object | null,
+  }> = (await Attainable.findAll({
+    where: {
+      courseId,
+      courseInstanceId,
+    },
+    attributes: [
+      'id',
+      'attainableId',
+      'formula',
+      'parentFormulaParams',
+    ],
+  })).map(
+    // TODO: Remove this map() call.
+    // map() is called as a temporary measure to translate attainableId to parentId.
+    (attainment: Attainable) => {
```

Comment:
If we decide to keep this map call or a similar construct, I would like to avoid having `Attainable` as the type here. Certain properties of `Attainable` are undefined, going against the interface of `Attainable`.